// ant-design-vue/es/menu/src/hooks/useMenuContext.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { defineComponent, inject, provide } from 'vue';
var MenuContextKey = Symbol('menuContextKey');

var useProvideMenu = function useProvideMenu(props) {
  provide(MenuContextKey, props);
};

var useInjectMenu = function useInjectMenu() {
  return inject(MenuContextKey);
};

var MenuFirstLevelContextKey = Symbol('menuFirstLevelContextKey');

var useProvideFirstLevel = function useProvideFirstLevel(firstLevel) {
  provide(MenuFirstLevelContextKey, firstLevel);
};

var useInjectFirstLevel = function useInjectFirstLevel() {
  return inject(MenuFirstLevelContextKey, true);
};

var MenuContextProvider = defineComponent({
  name: 'MenuContextProvider',
  inheritAttrs: false,
  props: {
    props: Object
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
    useProvideMenu(_extends(_extends({}, useInjectMenu()), props.props));
    return function () {
      var _a;

      return (_a = slots.default) === null || _a === void 0 ? void 0 : _a.call(slots);
    };
  }
});
export { useProvideMenu, MenuContextKey, useInjectMenu, MenuFirstLevelContextKey, useProvideFirstLevel, useInjectFirstLevel, MenuContextProvider };
export default useProvideMenu;
